class_name PrelimCombatHandler
extends Node

const PLAYER1_SCENE = preload("res://scenes/entities/player1.tscn")
const PLAYER2_SCENE = preload("res://scenes/entities/player2.tscn")
const ENEMY_SCENE = preload("res://scenes/entities/boss.tscn")
const MAX_PLAYERS: int = 2
const MAX_SKILL_SLOTS: int = 3

var entity_lib: EntityLibrary = null
var players: Array[Entity] = []
var enemy: Entity


func _ready()->void:
	entity_lib = EntityLibrary.new(get_parent().get_node('EntityData'))
	
	
func setup_fight() -> void:	
	spawn_players()
	spawn_enemy()
	#init_skills()  # Setup skills for all entities
	apply_skills_to_UI()


func spawn_players() -> void:
	#var player_positions = [Vector2(-450, -200), Vector2(-400, 150)]
	var player_positions = [Vector2(-600, -150), Vector2(-150, -150)]
	
	Logging.log('Spawning players')
	for i in range(MAX_PLAYERS):
		var player_instance: Entity
		if i == 0:
			player_instance = PLAYER1_SCENE.instantiate() as Entity
		else:
			player_instance = PLAYER2_SCENE.instantiate() as Entity

		player_instance.global_position = player_positions[i]
		#player_instance.name = "Player_%d" % (i + 1)
		
		entity_lib.load_data(player_instance)
		
		add_child(player_instance)
		players.append(player_instance)
		
		# update the player's ui (ui_handler) test
		UIHandler.refresh_ui(player_instance)
		
		#generate skills bar for this player
		var bar = player_instance.get_node_or_null("SkillsBar")
		if bar:
			bar.generate_placeholder_skill_icons()
			
		#generate skills column for this player
		var col = player_instance.get_node_or_null("SkillsColumn")
		if col:
			col.generate_placeholder_skill_icons()
			
		print((i + 1))

func spawn_enemy() -> void:
	var enemy_position = Vector2(200, 0)
	enemy_position = Vector2(600, -150)
	
	var enemy_instance = ENEMY_SCENE.instantiate() as Entity
	enemy_instance.global_position = enemy_position
	entity_lib.load_data(enemy_instance)
	#enemy_instance.name = "Enemy"
	
	add_child(enemy_instance)
	enemy = enemy_instance
	
	# update the boss's ui (ui_handler) test
	UIHandler.refresh_ui(enemy_instance)
	
	var bar: BossSkillsBar = enemy_instance.get_node_or_null("SkillsBar")
	if bar:
		bar.show_bar()
	else:
		push_error("Boss has NO SkillsBar node!")
		

#func init_skills() -> void:
	#
	## TODO: We should eventually switch to using a character library instead
	## of using this hardcoded template.
	#var skill_templates = {
		#1: Skill.new(1),
		#2: Skill.new(2),
		#3: Skill.new(3)
	#}
	#
	## Assigns skills to each player
	#for player in players:
		#player.skills = []
		#for skill_id in [1,1,1,1,1,1,2,2,2,2,3,3]:
			#var skill_instance = skill_templates[skill_id].duplicate()
			#player.skills.append(skill_instance)
		#
		## Max skills per turn
		#player.max_skill_slots = MAX_SKILL_SLOTS
	#
	## Assigns skills to the enemy
	#enemy.skills = []
	#for skill_id in [1,1,1,2,2,3]:
		#var skill_instance = skill_templates[skill_id].duplicate()
		#enemy.skills.append(skill_instance)
		
func apply_skills_to_UI():
	for player in players:
		var col := player.get_node_or_null("SkillsColumn")
		if col:
			col.show_skills(player)
